### PR TITLE
Generate F1 / IOU stats for USFIMR JRC experiments

### LINKED ISSUE #103: Generate IoUs for usfimr + jrc
New predictions based on the label-set of USFIMR + JRC surface water @jamesmcclain has produced for a few test cases require further evaluation. We can assist in this evaluation by generating IoU numbers (https://en.wikipedia.org/wiki/Jaccard_index).

The predictions live within the various folders' `predict` directory here: `s3://noaafloodmap-data-us-east-1/jmcclain/October_13_1307`. E.g. `s3://noaafloodmap-data-us-east-1/jmcclain/October_13_1307/SEN1FLOODS11_HAND/predict` and `s3://noaafloodmap-data-us-east-1/jmcclain/October_13_1307/USFIMR_TF_analyzed/predict`. The `TT`, `FT`, `FF`, and `TF` relate whether HAND data was used and whether the training was 3 class (`permanent water` + `flood water` + `other` as opposed to `water` + `other`) respectively.

In addition to calculating intersection and union per class, we'd like to break them down according to whether they are urban or not according to their categories from the latest NLCD (urban == 21 through 24 as laid out here: https://www.mrlc.gov/data/legends/national-land-cover-database-2016-nlcd2016-legend). We should already have NLCD under `s3://geotrellis-test/courage-services/nlcd`. It is likely simplest to get intersection and union values per these more granular categories and then sum them up for total IoUs without respect to NLCD categories

### PR DESCRIPTION
**Pull request by CloudNiner**
TODO:
- [x] Add ground truth uris for sen1floods11 experiments
- [x] Figure out what the additional params to the `f1_score` and `jaccard_score` method should be so that we only compute the score based on the water labels
- [x] Verify np mask logic
- [x] Run it and debug...

Closes #103

### CODE REVIEW

**Comment by CloudNiner** [comment]
Here's the CSV generated for SEN1FLOODS11_HAND from the code at 26d9b05e70852155611c44cda002052a11b9aaea:

[SEN1FLOODS11_HAND-iou-f1.csv.zip](https://github.com/azavea/noaa-flood-mapping/files/5378783/SEN1FLOODS11_HAND-iou-f1.csv.zip)

**Comment by CloudNiner** [comment]
@jamesmcclain @moradology this is ready for a look. Please use 🔍 because numpy masking and use of these functions is all new for me.

**Comment by jamesmcclain** [comment]
I have not run this, the logic looks okay to me. One thing that I noticed is that the "permanent water" and "flood water" classes in the ground-truth need to be collapsed into a single "water" class for the two-class models (because the two-class models only generate a single "water" prediction which encompasses both).

**Comment by CloudNiner** [comment]
> One thing that I noticed is that the "permanent water" and "flood water" classes in the ground-truth need to be collapsed into a single "water" class for the two-class models (because the two-class models only generate a single "water" prediction which encompasses both).

Can you clarify this? sen1floods11 and the two class USFIMR models use `labels=[1]` when calling into the stats methods and the source rasters only have one water label in them as far as I understand. The three class USFMIR uses `labels=[1,2]` in order to compute stats for each label independently. So I'm not sure what needs to be collapsed?

**Comment by jamesmcclain** [comment]
> > One thing that I noticed is that the "permanent water" and "flood water" classes in the ground-truth need to be collapsed into a single "water" class for the two-class models (because the two-class models only generate a single "water" prediction which encompasses both).
> 
> Can you clarify this? sen1floods11 and the two class USFIMR models use `labels=[1]` when calling into the stats methods and the source rasters only have one water label in them as far as I understand. The three class USFMIR uses `labels=[1,2]` in order to compute stats for each label independently. So I'm not sure what needs to be collapsed?

Yes, it is true that the predictions from the the two-class models only have one water class.  The ground-truth rasters have two water classes, though (1 and 2).  When comparing two-class predictions to the the (three-class) ground-truth labels, it is necessary to transform all 2s to 1s in the latter for the comparison to be valid.

**Comment by CloudNiner** [comment]
Ohhhhh. Gotcha. Update incoming.

**Comment by jamesmcclain** [comment]
> Ohhhhh. Gotcha. Update incoming.

I went back to get [this](https://github.com/azavea/noaa-flood-mapping/blob/07e9ac5b86e632a1ea6a14f4fbc143f5930ee22f/pipeline/usfimr_raster_pipeline.py#L121-L128) before I saw this response, but it might still be helpful.  It shows the transformation that is done in the training code.

**Comment by CloudNiner** [comment]
Here's the fix: https://github.com/azavea/noaa-flood-mapping/pull/104/commits/537a10769e8dc4d7658d39ed7767eb1815c2b475#diff-5f4c80a16f093bb0b7a0778743dff04ab6b943cd2754286ef26961bb315bbc83R192-R196

I'm going to let the script run to generate the stats csvs for each experiment. If you see any other issues LMK asap so I can cancel and address.

File: pipeline/usfimr-jrc-iou/main.py [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+                        transform=ds_p.transform,
+                        nodata=NODATA,
+                    )
+
+                    p_band = ds_p.read(1).flatten()
+                    t_band = ds_t.read(1).flatten()
+                    nlcd_band = nlcd_vrt.read(1).flatten()
+
+                assert len(p_band) == len(t_band)
+                assert len(p_band) == len(nlcd_band)
```
Comment (moradology) [inline pipeline/usfimr-jrc-iou/main.py:197]: Smart. I'm surprised this always works out via `WarpedVRT`

File: pipeline/usfimr-jrc-iou/main.py [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+
+                # TODO: Encode this in Experiment somehow...
+                # For two class USFIMR experiments, collapse ground truth three class
+                # labels for water to two class: 2 (flood) + 1 (perm) -> 1 (water)
+                if experiment.id.startswith("USFIMR") and len(experiment.labels) == 1:
+                    t_band = np.where(t_band == 2, 1, t_band)
```
Comment (moradology) [inline pipeline/usfimr-jrc-iou/main.py:203]: This is fine, but I think it might be more idiomatic to just say `t_band = t_band == 2`

File: pipeline/usfimr-jrc-iou/main.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+                nlcd_urban_mask = np.ma.masked_outside(nlcd_band, 21, 24).mask
+                p_band_urban = np.ma.array(p_band, mask=nlcd_urban_mask).filled(NODATA)
+                t_band_urban = np.ma.array(t_band, mask=nlcd_urban_mask).filled(NODATA)
```
Comment (moradology) [inline pipeline/usfimr-jrc-iou/main.py:207]: The masking API is pretty good here. New to me

**Review by moradology — APPROVED** [review]
As @jamesmcclain  said, the logic appears correct. Dropping to 1 dimension and using WarpedVRT were smart for this; relatively simple to think through what's going on

File: pipeline/usfimr-jrc-iou/main.py (hunk 2 above)
Comment (CloudNiner) [reply pipeline/usfimr-jrc-iou/main.py:203]: This was going to blow my mind, but it looks like the result is different for each operation. In this bit of code we want to replace all instances of `2` in the array with `1`:

```
In [1]: import numpy as np
In [2]: a = np.array([0, 1, 1, 2, 2])
In [3]: a == 2
Out[3]: array([False, False, False,  True,  True])
In [4]: np.where(a == 2, 1, a)
Out[4]: array([0, 1, 1, 1, 1])
```

File: pipeline/usfimr-jrc-iou/main.py (hunk 3 above)
Comment (CloudNiner) [reply pipeline/usfimr-jrc-iou/main.py:207]: Agreed. Worked well. In more complicated use cases i could see it being difficult keeping track of whether a var is a np.array or np.ma.masked_array, as the behavior with some np.array operations could lead to errors if you pass one where you intend to pass the other.

**Review by jamesmcclain — APPROVED** [review]
One nice-to-have would be overall f1 and IOU scores for all of the scenes, but I can mostly reconstruct those numbers from what is here plus image metadata.

**Comment by CloudNiner** [comment]
Discussed offline, @jamesmcclain is going to tweak the output for his purposes a bit later today. Merging this now.
